identity/src/main/java/androidx/security/identity/GattServer.java

```diff
@@ -184,33 +228,103 @@ public void onConnectionStateChange(BluetoothDevice device, int status, int newS
 
     @Override
     public void onCharacteristicReadRequest(BluetoothDevice device, int requestId, int offset,
-            BluetoothGattCharacteristic characteristic) {
+                                            BluetoothGattCharacteristic characteristic) {
         if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC) != 0) {
             Log.i(TAG, "onCharacteristicReadRequest: " + requestId + " " + offset + " " + characteristic.getUuid());
         }
         if (mCharacteristicIdentUuid != null
-            && characteristic.getUuid().equals(mCharacteristicIdentUuid)) {
+                && characteristic.getUuid().equals(mCharacteristicIdentUuid)) {
             mGattServer.sendResponse(device,
                     requestId,
                     BluetoothGatt.GATT_SUCCESS,
                     0,
                     mIdentValue);
+        } else if (mCharacteristicL2CAP != null
+                && characteristic.getUuid().equals(mCharacteristicL2CAPUuid)) {
+            if (mServerSocket == null) {
+                reportError(new Error("Socket for L2CAP not available"));
+            }
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC) != 0) {
+                    Log.i(TAG, "PSM Value generated: " + mServerSocket.getPsm() + " byte: " + (byte) mServerSocket.getPsm());
+                }
+                mGattServer.sendResponse(device,
+                        requestId,
+                        BluetoothGatt.GATT_SUCCESS,
+                        0,
+                        new byte[]{(byte) mServerSocket.getPsm()});
+                acceptConnection();
+            }
         } else {
             reportError(new Error("Read on unexpected characteristic with UUID "
                     + characteristic.getUuid()));
         }
     }
 
+    private void acceptConnection() {
+        try {
+            mSocket = mServerSocket.accept();
+            if (mSocket.isConnected()) {
+                // Stop accepting new connections
+                mServerSocket.close();
```

Comment:
Set `mServerSocket` to null here? Else we'll try to close it twice, which seems like it could result in errors.